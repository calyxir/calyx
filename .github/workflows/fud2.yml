name: fud2 tests

on:
  push:
  pull_request:
    branches: [main]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Install Rust
        uses: actions-rust-lang/setup-rust-toolchain@v1
      - run: cargo test -p fud
