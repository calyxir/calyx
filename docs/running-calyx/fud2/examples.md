# Examples

These commands will assume you're in the root directory of the Calyx
repository and that you have compiled Calyx.

**Simulating Calyx.**
Fud2 can compile a Calyx program to Verilog and simulate it using Verilator.

```bash
# Compile and simulate a vectorized add implementation
# in Calyx using the data provided,
# then dump the vcd into a new file for debugging.
# === Calyx:   examples/futil/vectorized-add.futil
# === data:    examples/dahlia/vectorized-add.fuse.data
# === output:  v-add.vcd
fud2 examples/futil/vectorized-add.futil \
  -o v-add.vcd \
  -s sim.data=examples/dahlia/vectorized-add.fuse.data \
```

**Simulating Dahlia.**
The following command prints out the final state of all memories by specifying
`--to dat`.

```bash
# Compile a Dahlia dot product implementation and
# simulate in verilog using the data provided.
# === Dahlia: examples/dahlia/dot-product.fuse
# === data:   examples/dahlia/dot-product.fuse.data
#     (`.data` is used as an extension alias for `.json`)
fud2 examples/dahlia/dot-product.fuse \
  --to dat \
  -s sim.data=examples/dahlia/dot-product.fuse.data
```

**Interpreting Calyx.**
In addition to Verilator, fud can execute Calyx programs using the [interpreter](../interpreter.md).

```bash
# Execute a Calyx program without compiling it,
# producing a JSON snapshot of the final state.
# === Calyx:   tests/correctness/while.futil
# === data:    tests/correctness/while.futil.data
fud2 tests/correctness/while.futil \
  --to dat \
  -s sim.data=tests/correctness/while.futil.data \
  --through cider
```