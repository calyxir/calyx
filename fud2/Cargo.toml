[package]
name = "fud"
version = "0.0.2"
edition.workspace = true
authors.workspace = true
license-file.workspace = true
repository.workspace = true
homepage.workspace = true

keywords = ["build-tool"]
readme = "README.md"
categories = ["build-tool"]
description = "Compiler driver for the Calyx infrastructure"

[dependencies]
fud-core = { path = "fud-core", version = "0.0.2" }
anyhow.workspace = true
manifest-dir-macros = "0.1"
include_dir = "0.7"

[lib]
name = "fud2"
path = "src/lib.rs"

[[bin]]
name = "fud2"
path = "src/main.rs"

[dev-dependencies]
insta = "1.36.0"
