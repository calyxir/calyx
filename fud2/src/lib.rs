use fud_core::{
    exec::{SetupRef, StateRef},
    run::{EmitResult, Emitter},
    DriverBuilder,
};

fn setup_calyx(
    bld: &mut DriverBuilder,
    verilog: StateRef,
) -> (StateRef, SetupRef) {
    let calyx = bld.state("calyx", &["futil"]);
    let calyx_setup = bld.setup("Calyx compiler", |e| {
        e.config_var("calyx-base", "calyx.base")?;
        e.config_var_or(
            "calyx-exe",
            "calyx.exe",
            "$calyx-base/target/debug/calyx",
        )?;
        e.rule(
            "calyx",
            "$calyx-exe -l $calyx-base -b $backend $args $in > $out",
        )?;
        e.rule(
            "calyx-pass",
            "$calyx-exe -l $calyx-base -p $pass $args $in > $out",
        )?;
        Ok(())
    });
    bld.op(
        "calyx-to-verilog",
        &[calyx_setup],
        calyx,
        verilog,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "verilog")?;
            Ok(())
        },
    );
    (calyx, calyx_setup)
}

fn setup_dahlia(
    bld: &mut DriverBuilder,
    calyx: StateRef,
) -> (StateRef, SetupRef) {
    let dahlia = bld.state("dahlia", &["fuse"]);
    let dahlia_setup = bld.setup("Dahlia compiler", |e| {
        e.config_var("dahlia-exe", "dahlia")?;
        e.rule(
            "dahlia-to-calyx",
            "$dahlia-exe -b calyx --lower -l error $in -o $out",
        )?;
        Ok(())
    });
    bld.rule(&[dahlia_setup], dahlia, calyx, "dahlia-to-calyx");
    (dahlia, dahlia_setup)
}

fn setup_mrxl(
    bld: &mut DriverBuilder,
    calyx: StateRef,
) -> (StateRef, SetupRef) {
    let mrxl = bld.state("mrxl", &["mrxl"]);
    let mrxl_setup = bld.setup("MrXL compiler", |e| {
        e.var("mrxl-exe", "mrxl")?;
        e.rule("mrxl-to-calyx", "$mrxl-exe $in > $out")?;
        Ok(())
    });
    bld.rule(&[mrxl_setup], mrxl, calyx, "mrxl-to-calyx");
    (mrxl, mrxl_setup)
}

pub fn build_driver(bld: &mut DriverBuilder) {
    // The verilog state
    let verilog = bld.state("verilog", &["sv", "v"]);
    // Calyx.
    let (calyx, calyx_setup) = setup_calyx(bld, verilog);
    // Dahlia.
    setup_dahlia(bld, calyx);
    // MrXL.
    setup_mrxl(bld, calyx);

    // Shared machinery for RTL simulators.
    let dat = bld.state("dat", &["json"]);
    let vcd = bld.state("vcd", &["vcd"]);
    let simulator = bld.state("sim", &["exe"]);
    let sim_setup = bld.setup("RTL simulation", |e| {
        // Data conversion to and from JSON.
        e.config_var_or("python", "python", "python3")?;
        e.rsrc("json-dat.py")?;
        e.rule("hex-data", "$python json-dat.py --from-json $in $out")?;
        e.rule("json-data", "$python json-dat.py --to-json $out $in")?;

        // The Verilog testbench.
        e.rsrc("tb.sv")?;

        // The input data file. `sim.data` is required.
        let data_name = e.config_val("sim.data")?;
        let data_path = e.external_path(data_name.as_ref());
        e.var("sim_data", data_path.as_str())?;

        // Produce the data directory.
        e.var("datadir", "sim_data")?;
        e.build_cmd(
            &["$datadir"],
            "hex-data",
            &["$sim_data"],
            &["json-dat.py"],
        )?;

        // Rule for simulation execution.
        e.rule(
            "sim-run",
            "./$bin +DATA=$datadir +CYCLE_LIMIT=$cycle-limit $args > $out",
        )?;

        // More shared configuration.
        e.config_var_or("cycle-limit", "sim.cycle_limit", "500000000")?;

        Ok(())
    });
    bld.op(
        "simulate",
        &[sim_setup],
        simulator,
        dat,
        |e, input, output| {
            e.build_cmd(&["sim.log"], "sim-run", &[input, "$datadir"], &[])?;
            e.arg("bin", input)?;
            e.arg("args", "+NOTRACE=1")?;
            e.build_cmd(
                &[output],
                "json-data",
                &["$datadir", "sim.log"],
                &["json-dat.py"],
            )?;
            Ok(())
        },
    );
    bld.op("trace", &[sim_setup], simulator, vcd, |e, input, output| {
        e.build_cmd(
            &["sim.log", output],
            "sim-run",
            &[input, "$datadir"],
            &[],
        )?;
        e.arg("bin", input)?;
        e.arg("args", &format!("+NOTRACE=0 +OUT={}", output))?;
        Ok(())
    });

    // Icarus Verilog.
    let verilog_noverify = bld.state("verilog-noverify", &["sv"]);
    let icarus_setup = bld.setup("Icarus Verilog", |e| {
        e.var("iverilog", "iverilog")?;
        e.rule("icarus-compile", "$iverilog -g2012 -o $out tb.sv $in")?;
        Ok(())
    });
    bld.op(
        "calyx-noverify",
        &[calyx_setup],
        calyx,
        verilog_noverify,
        |e, input, output| {
            // Icarus requires a special --disable-verify version of Calyx code.
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "verilog")?;
            e.arg("args", "--disable-verify")?;
            Ok(())
        },
    );
    bld.op(
        "icarus",
        &[sim_setup, icarus_setup],
        verilog_noverify,
        simulator,
        |e, input, output| {
            e.build_cmd(&[output], "icarus-compile", &[input], &["tb.sv"])?;
            Ok(())
        },
    );

    // Calyx to FIRRTL.
    let firrtl = bld.state("firrtl", &["fir"]);
    bld.op(
        "calyx-to-firrtl",
        &[calyx_setup],
        calyx,
        firrtl,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "firrtl")?;
            Ok(())
        },
    );

    // The FIRRTL compiler.
    let firrtl_setup = bld.setup("Firrtl to Verilog compiler", |e| {
        e.config_var("firrtl-exe", "firrtl.exe")?;
        e.rule("firrtl", "$firrtl-exe -i $in -o $out -X sverilog")?;

        e.rsrc("primitives-for-firrtl.sv")?;
        e.rule(
            "add-firrtl-prims",
            "cat primitives-for-firrtl.sv $in > $out",
        )?;

        Ok(())
    });
    fn firrtl_compile(
        e: &mut Emitter,
        input: &str,
        output: &str,
    ) -> EmitResult {
        let tmp_verilog = "partial.sv";
        e.build_cmd(&[tmp_verilog], "firrtl", &[input], &[])?;
        e.build_cmd(
            &[output],
            "add-firrtl-prims",
            &[tmp_verilog],
            &["primitives-for-firrtl.sv"],
        )?;
        Ok(())
    }
    bld.op("firrtl", &[firrtl_setup], firrtl, verilog, firrtl_compile);
    // This is a bit of a hack, but the Icarus-friendly "noverify" state is identical for this path
    // (since FIRRTL compilation doesn't come with verification).
    bld.op(
        "firrtl-noverify",
        &[firrtl_setup],
        firrtl,
        verilog_noverify,
        firrtl_compile,
    );

    // primitive-uses backend
    let primitive_uses_json = bld.state("primitive-uses-json", &["json"]);
    bld.op(
        "primitive-uses",
        &[calyx_setup],
        calyx,
        primitive_uses_json,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "primitive-uses")?;
            Ok(())
        },
    );

    // Verilator.
    let verilator_setup = bld.setup("Verilator", |e| {
        e.config_var_or("verilator", "verilator.exe", "verilator")?;
        e.config_var_or("cycle-limit", "sim.cycle_limit", "500000000")?;
        e.rule(
            "verilator-compile",
            "$verilator $in tb.sv --trace --binary --top-module TOP -fno-inline -Mdir $out-dir",
        )?;
        e.rule("cp", "cp $in $out")?;
        Ok(())
    });
    bld.op(
        "verilator",
        &[sim_setup, verilator_setup],
        verilog,
        simulator,
        |e, input, output| {
            let out_dir = "verilator-out";
            let sim_bin = format!("{}/VTOP", out_dir);
            e.build_cmd(
                &[&sim_bin],
                "verilator-compile",
                &[input],
                &["tb.sv"],
            )?;
            e.arg("out-dir", out_dir)?;
            e.build("cp", &sim_bin, output)?;
            Ok(())
        },
    );

    // Interpreter.
    let debug = bld.state("debug", &[]); // A pseudo-state.
    let cider_setup = bld.setup("Cider interpreter", |e| {
        e.config_var_or(
            "cider-exe",
            "cider.exe",
            "$calyx-base/target/debug/cider",
        )?;
        e.rule(
            "cider",
            "$cider-exe -l $calyx-base --raw --data data.json $in > $out",
        )?;
        e.rule(
            "cider-debug",
            "$cider-exe -l $calyx-base --data data.json $in debug || true",
        )?;
        e.arg("pool", "console")?;

        // TODO Can we reduce the duplication around and `$python`?
        e.rsrc("interp-dat.py")?;
        e.config_var_or("python", "python", "python3")?;
        e.rule("dat-to-interp", "$python interp-dat.py --to-interp $in")?;
        e.rule(
            "interp-to-dat",
            "$python interp-dat.py --from-interp $in $sim_data > $out",
        )?;
        e.build_cmd(
            &["data.json"],
            "dat-to-interp",
            &["$sim_data"],
            &["interp-dat.py"],
        )?;
        Ok(())
    });
    bld.op(
        "interp",
        &[sim_setup, calyx_setup, cider_setup],
        calyx,
        dat,
        |e, input, output| {
            let out_file = "interp_out.json";
            e.build_cmd(&[out_file], "cider", &[input], &["data.json"])?;
            e.build_cmd(
                &[output],
                "interp-to-dat",
                &[out_file],
                &["$sim_data", "interp-dat.py"],
            )?;
            Ok(())
        },
    );
    bld.op(
        "debug",
        &[sim_setup, calyx_setup, cider_setup],
        calyx,
        debug,
        |e, input, output| {
            e.build_cmd(&[output], "cider-debug", &[input], &["data.json"])?;
            Ok(())
        },
    );

    // Xilinx compilation.
    let xo = bld.state("xo", &["xo"]);
    let xclbin = bld.state("xclbin", &["xclbin"]);
    let xilinx_setup = bld.setup("Xilinx tools", |e| {
        // Locations for Vivado and Vitis installations.
        e.config_var("vivado-dir", "xilinx.vivado")?;
        e.config_var("vitis-dir", "xilinx.vitis")?;

        // Package a Verilog program as an `.xo` file.
        e.rsrc("gen_xo.tcl")?;
        e.rsrc("get-ports.py")?;
        e.config_var_or("python", "python", "python3")?;
        e.rule("gen-xo", "$vivado-dir/bin/vivado -mode batch -source gen_xo.tcl -tclargs $out `$python get-ports.py kernel.xml`")?;
        e.arg("pool", "console")?;  // Lets Ninja stream the tool output "live."

        // Compile an `.xo` file to an `.xclbin` file, which is where the actual EDA work occurs.
        e.config_var_or("xilinx-mode", "xilinx.mode", "hw_emu")?;
        e.config_var_or("platform", "xilinx.device", "xilinx_u50_gen3x16_xdma_201920_3")?;
        e.rule("compile-xclbin", "$vitis-dir/bin/v++ -g -t $xilinx-mode --platform $platform --save-temps --profile.data all:all:all --profile.exec all:all:all -lo $out $in")?;
        e.arg("pool", "console")?;

        Ok(())
    });
    bld.op(
        "xo",
        &[calyx_setup, xilinx_setup],
        calyx,
        xo,
        |e, input, output| {
            // Emit the Verilog itself in "synthesis mode."
            e.build_cmd(&["main.sv"], "calyx", &[input], &[])?;
            e.arg("backend", "verilog")?;
            e.arg("args", "--synthesis -p external")?;

            // Extra ingredients for the `.xo` package.
            e.build_cmd(&["toplevel.v"], "calyx", &[input], &[])?;
            e.arg("backend", "xilinx")?;
            e.build_cmd(&["kernel.xml"], "calyx", &[input], &[])?;
            e.arg("backend", "xilinx-xml")?;

            // Package the `.xo`.
            e.build_cmd(
                &[output],
                "gen-xo",
                &[],
                &[
                    "main.sv",
                    "toplevel.v",
                    "kernel.xml",
                    "gen_xo.tcl",
                    "get-ports.py",
                ],
            )?;
            Ok(())
        },
    );
    bld.op("xclbin", &[xilinx_setup], xo, xclbin, |e, input, output| {
        e.build_cmd(&[output], "compile-xclbin", &[input], &[])?;
        Ok(())
    });

    // Xilinx execution.
    // TODO Only does `hw_emu` for now...
    let xrt_setup = bld.setup("Xilinx execution via XRT", |e| {
        // Generate `emconfig.json`.
        e.rule("emconfig", "$vitis-dir/bin/emconfigutil --platform $platform")?;
        e.build_cmd(&["emconfig.json"], "emconfig", &[], &[])?;

        // Execute via the `xclrun` tool.
        e.config_var("xrt-dir", "xilinx.xrt")?;
        e.rule("xclrun", "bash -c 'source $vitis-dir/settings64.sh ; source $xrt-dir/setup.sh ; XRT_INI_PATH=$xrt_ini EMCONFIG_PATH=. XCL_EMULATION_MODE=$xilinx-mode $python -m fud.xclrun --out $out $in'")?;
        e.arg("pool", "console")?;

        // "Pre-sim" and "post-sim" scripts for simulation.
        e.rule("echo", "echo $contents > $out")?;
        e.build_cmd(&["pre_sim.tcl"], "echo", &[""], &[""])?;
        e.arg("contents", "open_vcd\\\\nlog_vcd *\\\\n")?;
        e.build_cmd(&["post_sim.tcl"], "echo", &[""], &[""])?;
        e.arg("contents", "close_vcd\\\\n")?;

        Ok(())
    });
    bld.op(
        "xrt",
        &[xilinx_setup, sim_setup, xrt_setup],
        xclbin,
        dat,
        |e, input, output| {
            e.rsrc("xrt.ini")?;
            e.build_cmd(
                &[output],
                "xclrun",
                &[input, "$sim_data"],
                &["emconfig.json", "xrt.ini"],
            )?;
            e.arg("xrt_ini", "xrt.ini")?;
            Ok(())
        },
    );
    bld.op(
        "xrt-trace",
        &[xilinx_setup, sim_setup, xrt_setup],
        xclbin,
        vcd,
        |e, input, output| {
            e.rsrc("xrt_trace.ini")?;
            e.build_cmd(
                &[output], // TODO not the VCD, yet...
                "xclrun",
                &[input, "$sim_data"],
                &[
                    "emconfig.json",
                    "pre_sim.tcl",
                    "post_sim.tcl",
                    "xrt_trace.ini",
                ],
            )?;
            e.arg("xrt_ini", "xrt_trace.ini")?;
            Ok(())
        },
    );

    let yxi = bld.state("yxi", &["yxi"]);
    bld.op(
        "calyx-to-yxi",
        &[calyx_setup],
        calyx,
        yxi,
        |e, input, output| {
            e.build_cmd(&[output], "calyx", &[input], &[])?;
            e.arg("backend", "yxi")?;
            Ok(())
        },
    );

    let wrapper_setup = bld.setup("YXI and AXI generation", |e| {
        // Define a `gen-axi` rule that invokes our Python code generator program.
        // For now point to standalone axi-generator.py. Can maybe turn this into a rsrc file?
        e.config_var_or(
            "axi-generator",
            "axi.generator",
            "$calyx-base/yxi/axi-calyx/axi-generator.py",
        )?;
        e.config_var_or("python", "python", "python3")?;
        e.rule("gen-axi", "$python $axi-generator $in > $out")?;

        // Define a simple `combine` rule that just concatenates any numer of files.
        e.rule("combine", "cat $in > $out")?;

        e.rule(
            "remove-imports",
            "sed '1,/component main/{/component main/!d}' $in > $out",
        )?;
        Ok(())
    });
    bld.op(
        "axi-wrapped",
        &[calyx_setup, wrapper_setup],
        calyx,
        calyx,
        |e, input, output| {
            // Generate the YXI file.
            //no extension
            let file_name = input
                .rsplit_once('/')
                .unwrap()
                .1
                .rsplit_once('.')
                .unwrap()
                .0;
            let tmp_yxi = format!("{}.yxi", file_name);

            //Get yxi file from main compute program.
            //TODO(nate): Can this use the `yxi` operation instead of hardcoding the build cmd calyx rule with arguments?
            e.build_cmd(&[&tmp_yxi], "calyx", &[input], &[])?;
            e.arg("backend", "yxi")?;

            // Generate the AXI wrapper.
            let refified_calyx = format!("refified_{}.futil", file_name);
            e.build_cmd(&[&refified_calyx], "calyx-pass", &[input], &[])?;
            e.arg("pass", "external-to-ref")?;

            let axi_wrapper = "axi_wrapper.futil";
            e.build_cmd(&[axi_wrapper], "gen-axi", &[&tmp_yxi], &[])?;

            // Generate no-imports version of the refified calyx.
            let no_imports_calyx = format!("no_imports_{}", refified_calyx);
            e.build_cmd(
                &[&no_imports_calyx],
                "remove-imports",
                &[&refified_calyx],
                &[],
            )?;

            // Combine the original Calyx and the wrapper.
            e.build_cmd(
                &[output],
                "combine",
                &[axi_wrapper, &no_imports_calyx],
                &[],
            )?;
            Ok(())
        },
    );
}
