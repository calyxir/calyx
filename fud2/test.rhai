
let dahlia2 = state("dahlia2", ["fuse"]);
let calyx2 = state("calyx2", ["futil"]);

fn dahlia_setup(e) {
   e.config_var("dahlia-exe", "dahlia");
   let test = e.config_val("sim.data");
   e.comment(test); 
   e.rule("dahlia2-to-calyx2", "$dahlia-exe -b calyx --lower -l error $in -o $out");
}

rule([dahlia_setup], dahlia2, calyx2, "dahlia2-to-calyx2");
