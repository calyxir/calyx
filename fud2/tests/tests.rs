use fud2::build_driver;
use fud_core::{
    config::default_config, exec::Request, run::Run, Driver, DriverBuilder,
};

fn test_driver() -> Driver {
    let mut bld = DriverBuilder::new("fud2");
    build_driver(&mut bld);
    bld.build()
}

fn request(
    driver: &Driver,
    start: &str,
    end: &str,
    through: &[&str],
) -> Request {
    fud_core::exec::Request {
        start_file: None,
        start_state: driver.get_state(start).unwrap(),
        end_file: None,
        end_state: driver.get_state(end).unwrap(),
        through: through.iter().map(|s| driver.get_op(s).unwrap()).collect(),
        workdir: ".".into(),
    }
}

fn emit_ninja(driver: &Driver, req: Request) -> String {
    let plan = driver.plan(req).unwrap();
    let config = default_config()
        .merge(("exe", "fud2"))
        .merge(("calyx.base", "/test/calyx"))
        .merge(("firrtl.exe", "/test/bin/firrtl"))
        .merge(("sim.data", "/test/data.json"))
        .merge(("xilinx.vivado", "/test/xilinx/vivado"))
        .merge(("xilinx.vitis", "/test/xilinx/vitis"))
        .merge(("xilinx.xrt", "/test/xilinx/xrt"))
        .merge(("dahlia", "/test/bin/dahlia"));
    let run = Run::with_config(driver, plan, config);
    let mut buf = vec![];
    run.emit(&mut buf).unwrap();
    String::from_utf8(buf).unwrap()
}

/// Get a human-readable description of a request.
fn req_desc(driver: &Driver, req: &Request) -> String {
    let mut desc = format!(
        "emit {} -> {}",
        driver.states[req.start_state].name, driver.states[req.end_state].name
    );
    if !req.through.is_empty() {
        desc.push_str(" through");
        for op in &req.through {
            desc.push(' ');
            desc.push_str(&driver.ops[*op].name);
        }
    }
    desc
}

/// Get a short string uniquely identifying a request.
fn req_slug(driver: &Driver, req: &Request) -> String {
    let mut desc = driver.states[req.start_state].name.to_string();
    for op in &req.through {
        desc.push('_');
        desc.push_str(&driver.ops[*op].name);
    }
    desc.push('_');
    desc.push_str(&driver.states[req.end_state].name);
    desc
}

fn test_emit(driver: &Driver, req: Request) {
    let desc = req_desc(driver, &req);
    let slug = req_slug(driver, &req);
    let ninja = emit_ninja(driver, req);
    insta::with_settings!({
        description => desc,
        omit_expression => true,
        snapshot_suffix => slug,
    }, {
        insta::assert_snapshot!(ninja);
    });
}

#[test]
fn calyx_to_verilog() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "verilog", &[]));
}

#[test]
fn calyx_via_firrtl() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "verilog", &["firrtl"]));
}

#[test]
fn sim_tests() {
    let driver = test_driver();
    for dest in &["dat", "vcd"] {
        for sim in &["icarus", "verilator"] {
            test_emit(&driver, request(&driver, "calyx", dest, &[sim]));
        }
    }
}

#[test]
fn cider_tests() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "dat", &["interp"]));
    test_emit(&driver, request(&driver, "calyx", "debug", &[]));
}

#[test]
fn xrt_tests() {
    let driver = test_driver();
    test_emit(&driver, request(&driver, "calyx", "dat", &["xrt"]));
    test_emit(&driver, request(&driver, "calyx", "vcd", &["xrt-trace"]));
}

#[test]
fn frontend_tests() {
    let driver = test_driver();
    for frontend in &["dahlia", "mrxl"] {
        test_emit(&driver, request(&driver, frontend, "calyx", &[]));
    }
}
