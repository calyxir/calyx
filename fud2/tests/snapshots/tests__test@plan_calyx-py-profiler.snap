---
source: fud2/tests/tests.rs
description: "emit plan: calyx-py-profiler"
---
build-tool = fud2
rule get-rsrc
  command = $build-tool get-rsrc $out

calyx-base = /test/calyx
calyx-exe = $calyx-base/target/debug/calyx
calyx-lib-path = $calyx-base
args = 
rule calyx
  command = $calyx-exe -l $calyx-lib-path -b $backend $args $in > $out
rule calyx-pass
  command = $calyx-exe -l $calyx-lib-path -p $pass $args $in > $out
cider-calyx-passes = -p none
rule calyx-cider
  command = $calyx-exe -l $calyx-lib-path $cider-calyx-passes $args $in > $out

cells = cells.json
fsm-json = fsm.json
cell-share-json = shared-cells.json
passes = -p pre-opt -p compile -p post-opt -p lower
edsl-args = 
component_cells = $calyx-base/target/debug/component_cells
rule component-cells
  command = $component_cells -l $calyx-base $in > $out
parse-vcd-script = $calyx-base/tools/profiler/profiler-process/main.py
rule parse-vcd
  command = python3 $parse-vcd-script $in $cells $fsm-json $cell-share-json profiler-out $out --print-trace-threshold 100
flamegraph-script = /test/calyx/non-existent.script
create-visuals-script = $calyx-base/tools/profiler/create-visuals.sh
rule create-visuals
  command = bash $create-visuals-script $flamegraph-script profiler-out $in $out
build tb.sv: get-rsrc
rule edsl-to-calyx
  command = python3 $in $edsl-args > $out
metadata-mapping-json = metadata-map.json
fileinfo_emitter = $calyx-base/target/debug/fileinfo_emitter
rule parse-metadata
  command = $fileinfo_emitter -l $calyx-base $in > $metadata-mapping-json
rule parse-vcd-from-adl
  command = python3 $parse-vcd-script $in $cells $fsm-json $cell-share-json profiler-out $out --adl-mapping-file $metadata-mapping-json --print-trace-threshold 100

verilator = verilator
cycle-limit = 500000000
rule verilator-compile-standalone-tb
  command = $verilator $in tb.sv --trace --binary --top-module toplevel -fno-inline -Mdir $out-dir
rule verilator-compile-custom-tb
  command = $verilator $in tb.sv memories.sv --trace --binary --top-module toplevel -fno-inline -Mdir $out-dir
rule cp
  command = cp $in $out

python = python3
build json-dat.py: get-rsrc
rule hex-data
  command = $python json-dat.py --from-json $in $out
rule json-data
  command = $python json-dat.py --to-json $out $in
sim_data = /test/data.json
datadir = sim_data
build $datadir: hex-data $sim_data | json-dat.py
rule sim-run
  command = ./$bin +DATA=$datadir +CYCLE_LIMIT=$cycle-limit $args > $out
cycle-limit = 500000000

build calyx.futil: edsl-to-calyx /input.ext
build $metadata-mapping-json: parse-metadata calyx.futil
build $cells: component-cells calyx.futil
build instrumented.sv: calyx calyx.futil
  backend = verilog
  args = -p validate -p compile-invoke -p profiler-instrumentation $passes -x tdcc:dump-fsm-json=fsm.json -x cell-share:emit-share-map=shared-cells.json
build verilator-out/Vtoplevel: verilator-compile-standalone-tb instrumented.sv | tb.sv
  out-dir = verilator-out
build instrumented.exe: cp verilator-out/Vtoplevel
build sim.log instrumented.vcd: sim-run instrumented.exe $datadir
  bin = instrumented.exe
  args = +NOTRACE=0 +OUT=instrumented.vcd
build flamegraph.folded: parse-vcd-from-adl instrumented.vcd | $cells $metadata-mapping-json
build /output.ext: create-visuals flamegraph.folded

default /output.ext
